src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+    /home/dan1dr/.cache/pypoetry/virtualenvs/zrive-ds-UEx3J_CK-py3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
+      _warn_prf(average, modifier, msg_start, len(result))
+
+
+                  precision    recall  f1-score   support
+    
+               0       0.99      1.00      0.99    213182
+               1       0.00      0.00      0.00      3213
+    
+        accuracy                           0.99    216395
+       macro avg       0.49      0.50      0.50    216395
+    weighted avg       0.97      0.99      0.98    216395
+    
+    Confusion Matrix:
+    [[213182      0]
+     [  3213      0]]
+
+
+    /home/dan1dr/.cache/pypoetry/virtualenvs/zrive-ds-UEx3J_CK-py3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
+      _warn_prf(average, modifier, msg_start, len(result))
+
+
+
+```python
+# Calculate probabilities for the positive class
+y_val_probs = model.predict_proba(X_val)[:, 1]
+
+# ROC Curve
+fpr_val, tpr_val, _ = roc_curve(y_val, y_val_probs)
+roc_auc_val = auc(fpr_val, tpr_val)  # Calculate AUC
+
+plt.figure(figsize=(10,5))
+
+plt.subplot(1, 2, 1)
+plt.plot(fpr_val, tpr_val, label=f'ROC (AUC = {roc_auc_val:.2f})')
+plt.plot([0, 1], [0, 1], color='navy', linestyle='--')
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('ROC')
+plt.legend(loc="lower right")
+
+
+# Precision-Recall Curve
+precision_val, recall_val, _ = precision_recall_curve(y_val, y_val_probs)
+
+plt.subplot(1, 2, 2)
+plt.plot(recall_val, precision_val)
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Precision-Recall Curve')
+
+plt.tight_layout()
+plt.show()
+```
+
+
+    
+![png](push_exploration_files/push_exploration_37_0.png)
+    
+
+
+We see the model it is very good for predicting the negative (no-purchase), but poor for positive (purchases). This may come from the strongly unbalanced classes we had in the dataset.
+
+
+```python
+final_data['outcome'].value_counts()
+```
+
+
+
+
+    outcome
+    0    2132624
+    1      31329
+    Name: count, dtype: int64
+
+
+
+
+```python
+final_data['any_event'].value_counts()
+```
+
+
+
+
+    any_event
+    0    2104207
+    1      59746
+    Name: count, dtype: int64
+
+
+
+First, let's usee the weight_class that scikit offers. Then we could think of oversampling the minority class, or undersampling the majority. We have 2M rows, so undersampling might be a good idea since we might have enough data for doing that. Oversampling may disrupt the data and add new ones, but may work also
```

Comment:
a couple of comments: 
* using weight_class should be equivalent to over / undersampling (depending on whether the weight is >1 or <1). 
* in my experience, this methods to "fix" imbalance don't work most of the time (or never), so I'm normally very skeptical about using them and try to avoid them